[Impeller] Make Rect origin and size fields private

flar opened the pull request:
Make the implementation fields of Rect private in preparation for switching its internal operations from XYWH to LTRB fields.

Additionally `[[nodiscard]]` tags are added to all of the Rect methods that return a value to avoid problems with confusing mutator methods and return value methods.

flar commented:
The new additions to the Rect class all have unit tests.

The changes in the rest of the file should be simply replacing an old reference with its replacement. More than 90% of the time `origin` -> `GetOrigin()` and `size` -> `GetSize()` but a couple of the usages were converted into new methods on the Rect class combined with tests that verify that the new methods do the same thing as the manual code they are replacing.

Please evaluate as part of your reviews if you think there are additional tests needed to double-check the substitutions.

dnfield commented on impeller/geometry/rect.h:
```diff
@@ -190,15 +207,44 @@ struct TRect {
+    return {origin.x + size.width * tx, origin.y + size.height * ty};
+  }
+
+  /// @brief  Get the pointe related to this rectangle associated with
```
```suggestion
  /// @brief  Get the point related to this rectangle associated with
```

flar replied:
Code deleted...

dnfield commented on impeller/geometry/rect.h:
```diff
@@ -190,15 +207,44 @@ struct TRect {
     return Point(origin.x + size.width * 0.5f, origin.y + size.height * 0.5f);
   }
 
+  /// @brief  Get the point related to this rectangle associated with
+  ///         the given fractional values |tx| and |ty| with a fractional
+  ///         value of 0 referring to the left or upper edge of the
+  ///         rectangle and a fractional value of 1 referring to the right
+  ///         or bottom edge of the rectangle.
+  ///
+  ///         GetRelative(0, 0) returns the upper left corner
+  ///         GetRelative(0.5, 0.5) returns the center of the rectangle
+  ///         GetRelative(1, 1) returns the lower right corner
+  constexpr Point GetRelative(Scalar tx, Scalar ty) const {
+    return {origin.x + size.width * tx, origin.y + size.height * ty};
+  }
+
+  /// @brief  Get the pointe related to this rectangle associated with
+  ///         the fractional values in the point |t|, equivalent to
+  ///         |GetRelative(t.x, t.y)|
+  constexpr Point GetRelative(const Point& t) const {
+    return GetRelative(t.x, t.y);
+  }
```
When reading the code using these methods it wasn't immediately apparent to me what they actually do. 

I wonder if there's some clearer name we could use? Looking at the Dart Rect and SkRect classes they don't quite have a convenience for this (though the Dart API has things like `rect.topCenter`).

I also wonder if we should DCHECK that `0 <= t <= 1`, although maybe that can't be constexpr anymore...

flar replied:
I didn't see a need for the [0,1] restriction, though. What if someone wants to find the point that is halfway to the left of the rectangle?

I agree the name isn't great and would love to see some suggestions.

flar replied:
Another thought - the code that uses this was only using 0,1 values and doing it to choose which corner was being inserted in a triangle list. That code could be rewritten to use a triangle strip and getPoints (which I believe is ordered for making a triangle strip) and we could avoid this new method entirely. But, I'd have to make sure that change was tested.

The code is in atlas_contents.cc seen modified in this PR...

flar replied:
New methods deleted in favor of a simpler implementation in atlas_contents.cc using existing methods.

dnfield reviewed (approved):
LGTM with a nitpick about GetRelative and a typo

flar commented:
The goldens have identified a bug somewhere. Looking into it...

flar commented:
> The goldens have identified a bug somewhere. Looking into it...

There were a couple of places where I called a Rect method without storing its return value. I've fixed those and added `[[nodiscard]]` directives to as many of the Rect methods as I thought would benefit from it in order to prevent future mistakes.

Before my first commit on this PR a unit test had already caught one other case where I was ignoring a return value so this scenario is likely to happen again if not for the no-discard tags just added...
